src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
+    colind = index(df)[indexcols]
+
+    if length(colind) == 0
+        throw(ArgumentError("At least one column to complete combinations required"))
+    end
+
+    has_duplicates = row_group_slots(ntuple(i -> df[!, colind[i]], length(colind)),
+                                     Val(false), nothing, false, nothing)[1] != nrow(df)
+    if has_duplicates && !allowduplicates
+        throw(ArgumentError("duplicate rows in input"))
+    end
+
+    # Create a vector of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        # All levels are retained, missing is added only if present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[!, col]), missing)
+        else
+            tempcol = levels(df[!, col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # make sure we do not overflow in the target data frame size
+    target_rows = Int(foldl((x, y) -> x * length(y), uniqueVals, init=big(1)))
+    if iszero(target_rows)
+        @assert iszero(nrow(df))
+        return allcols ? select(df, colind, :) : select(df, colind)
+    end
+
+    # construct expanded columns
+    out_df = DataFrame()
+    inner = 1
+    @assert length(uniqueVals) == length(colind)
+    for (val, cind) in zip(uniqueVals, colind)
+        len = length(val)
+        target_col = Tables.allocatecolumn(eltype(df[!, cind]), len)
+        target_col .= val
+        last_inner = inner
+        inner *= len
+        outer, remv = divrem(target_rows, inner)
+        @assert iszero(remv)
+        out_df[!, _names(df)[cind]] = repeat(target_col, inner=last_inner, outer=outer)
+    end
+    @assert inner == target_rows
+
+    # optionally join the remaining columns
+    if allcols
+        idx_ind = 0
+        while columnindex(df, string("source", idx_ind)) != 0
+            idx_ind += 1
+        end
+
+        if has_duplicates
+            order_ind = 0
+            while columnindex(df, string("order", order_ind)) != 0
+                order_ind += 1
+            end
+            insertcols!(out_df, 1, string("order", order_ind) => 1:nrow(out_df))
+            out_df = leftjoin(out_df, df; on=_names(df)[colind],
+                            source=string("source", idx_ind),
+                            matchmissing=:equal)
+            sort!(out_df, 1)
+            select!(out_df, 2:ncol(out_df))
+        else
+            leftjoin!(out_df, df; on=_names(df)[colind],
+                      source=string("source", idx_ind),
+                      matchmissing=:equal)
+        end
+        # Replace missing values with the fill
+        if !ismissing(fill)
+            mask = out_df[!, end] .== "left_only"
+            if count(mask) > 0
+                for n in length(colind)+1:ncol(out_df)-1
+                    tmp_col = out_df[!, n]
+                    out_col = similar(tmp_col,
```

Comment:
@nalimilan - this is needed to have consistent behavior with `unstack`, see also https://github.com/JuliaData/CategoricalArrays.jl/issues/380.